app/src/main/java/com/weatherxm/util/NumberUtils.kt

```diff
@@ -1,35 +1,61 @@
 package com.weatherxm.util
 
 import android.icu.text.CompactDecimalFormat
-import android.icu.text.NumberFormat
-import com.weatherxm.util.Weather.EMPTY_VALUE
+import com.weatherxm.data.DECIMAL_FORMAT_TOKENS
+import com.weatherxm.ui.common.Contracts.EMPTY_VALUE
 import org.koin.core.component.KoinComponent
 import org.koin.core.component.inject
+import org.koin.core.qualifier.named
+import java.math.BigDecimal
 import java.math.RoundingMode
+import java.text.DecimalFormat
+import java.text.NumberFormat
 
 object NumberUtils : KoinComponent {
     private val compactDecimalFormat: CompactDecimalFormat by inject()
     private val numberFormat: NumberFormat by inject()
+    private val decimalFormat: DecimalFormat by inject(named(DECIMAL_FORMAT_TOKENS))
+
+    private const val DIVISOR_WEI_TO_ETH = "1000000000000000000"
+    private const val ETH_DECIMALS = 18
 
     fun compactNumber(number: Number?): String {
         return number?.let {
             compactDecimalFormat.format(number)
         } ?: EMPTY_VALUE
     }
 
-    fun formatNumber(number: Number?, maxDecimals: Int = 0): String {
+    fun formatNumber(number: Number?, decimals: Int = 0): String {
         return number?.let {
-            numberFormat.maximumFractionDigits = maxDecimals
+            numberFormat.minimumFractionDigits = decimals
+            numberFormat.maximumFractionDigits = decimals
             numberFormat.format(it)
         } ?: EMPTY_VALUE
     }
 
-    fun roundToDecimals(value: Number, decimals: Int = 1): Float {
-        return value.toFloat().toBigDecimal().setScale(decimals, RoundingMode.HALF_UP).toFloat()
+    fun formatTokens(amount: Float?): String {
+        return amount?.let {
+            formatTokens(it.toBigDecimal())
+        } ?: EMPTY_VALUE
+    }
+
+    fun formatTokens(amount: BigDecimal): String {
+        return decimalFormat.format(amount)
+    }
+
+    @Suppress("MagicNumber")
+    fun weiToETH(amount: BigDecimal): BigDecimal {
+        /**
+         * Mandatory otherwise we get a result of 0E-18 instead of 0
+         */
+        if (amount == BigDecimal.ZERO) {
+            return BigDecimal.ZERO
+        }
+        return amount.divide(BigDecimal(DIVISOR_WEI_TO_ETH), ETH_DECIMALS, RoundingMode.HALF_UP)
     }
 
-    fun roundToInt(value: Number): Int {
-        return value.toFloat().toBigDecimal().setScale(0, RoundingMode.HALF_UP).toInt()
+    fun roundToDecimals(value: Number, decimals: Int = 1): Float {
```

Comment:
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2391143597>#b921d3</a><span> - 10/03/2024, 11:34 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The roundToDecimals function lacks input validation for the 'decimals' parameter, which could lead to unexpected behavior or errors if negative or excessively large values are provided.
</div>
<div id="fix">
<b>Fix:</b> Add input validation for the 'decimals' parameter to ensure it's non-negative and within a reasonable range.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -57,2 +57,7 @@ object NumberUtils : KoinComponent {
 -    fun roundToDecimals(value: Number, decimals: Int = 1): Float {
 -        return value.toFloat().toBigDecimal().setScale(decimals, RoundingMode.HALF_UP).toFloat()
 +    fun roundToDecimals(value: Number, decimals: Int = 1): Float {
 +        require(decimals >= 0) { "Decimals must be non-negative" }
 +        require(decimals <= 10) { "Decimals must be 10 or less" }
 +        return value.toFloat().toBigDecimal().setScale(decimals, RoundingMode.HALF_UP).toFloat()
 +    }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged